File: CrackingTheCodeInterview/Array&Strings/C++/StringCompression.cpp
#include <bits/stdc++.h>
#if !defined(True)
#define True  (1==1)
#endif
#if !defined(False)
#define False (!True)
#endif
#if !defined(None)
#define None  NULL
#endif
using namespace std;
typedef vector<char> vc;

void pushDigits(vc &str, size_t integer){
	string res, digits;							// O(1)
	ostringstream os;							// O(1)
	os << integer;								// O(1)
	digits = os.str();							// O(1) <- Max number handle by unsigned int
	for(auto chr:digits) str.push_back(chr);	// O(1) <- Max number handle by unsigned int
}

string stringCompression(string str){
	vc newString;                               // O(1)
	char auxChar = 1;                           // O(1)
	size_t count = 1;							// O(1)
	for(char chr: str)							// O(n)
		if(auxChar == 1) auxChar = chr;				// O(1)	
		else if(chr != auxChar){					// O(1)
			newString.push_back(auxChar);				// O(1)
			pushDigits(newString, count);				// O(1)
			auxChar = chr;								// O(1)
			count =  1;									// O(1)
		}else count++;								// O(1)
	newString.push_back(auxChar);			// O(1)
	pushDigits(newString, count);			// O(1)
	string res (newString.begin(), newString.end());
	return (res.length() >= str.length())? str : res;	// O(1)
}

int main(int argc, char const *argv[]){
	ios_base::sync_with_stdio(False);
	cout << stringCompression("aabcccccaaa") << std::endl;		
	cout << stringCompression("abbbbbbbbbbbb") << std::endl;		
	return 0;
}

File: CrackingTheCodeInterview/Array&Strings/Python3/StringCompression.py
def stringCompression(string):
	newString, auxChar, count = [], None, 1									# O(1)
	for char in string:														# O(n)
		if char != auxChar or auxChar is None:									# O(1)
			auxChar, count = char, 1												# O(1)
			newString.append(auxChar)												# O(1)
			newString.append('1')														# O(1)
		else:																	# O(1)
			count += 1																# O(1)
			newString[-1] = str(count)												# O(1)
	return string if len(newString) >= len(string) else ''.join(newString)	# O(1)

def stringCompressionList(string):
	auxChar, count = None, 1																# O(1)
	for index, char in enumerate(string):													# O(n)
		if char != auxChar or auxChar is None:													# O(1)
			auxChar, count = char, 1																# O(1)
		else:																					# O(1)
			count += 1																				# O(1)
			string[index] = count																	# O(1)
	for i in range(len(string) - 1):														# O(n)
		if isinstance(string[i], int) and isinstance(string[i + 1], int): string[i] = '_'		# O(1)
	for i in range(len(string)):															# O(n)
		if isinstance(string[i], int): string[i] = str(string[i])								# O(1)
	return ''.join(string).replace('_', '')
print(stringCompressionList(["a","a","b","b","c","c","c"]))
print(stringCompression('aabcccccaaa'))